pkg/controllers/provisioning/v1alpha1/reallocator/filter.go

```diff
@@ -25,40 +30,81 @@ type Filter struct {
 	kubeClient client.Client
 }
 
-// Gets Nodes fitting some underutilization
-// TODO: add predicate
-func (f *Filter) GetUnderutilizedNodes(ctx context.Context) ([]*v1.Node, error) {
-	nodeList := &v1.NodeList{}
+// Gets Nodes fitting some underutilization predicate
+func (f *Filter) GetUnderutilizedNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
 	underutilized := []*v1.Node{}
 
-	// 1. Get all nodes
-	if err := f.kubeClient.List(ctx, nodeList); err != nil {
-		return nil, fmt.Errorf("listing nodes, %w", err)
+	// 1. Get all provisioner labeled nodes
+	nodes, err := f.getNodes(ctx, provisioner)
+	if err != nil {
+		return nil, err
 	}
 
 	// 2. Get nodes and the pods on each node
-	for _, node := range nodeList.Items {
+	for _, node := range nodes {
 		pods, err := f.getPodsOnNode(ctx, node.Name)
 		if err != nil {
-			return []*v1.Node{}, fmt.Errorf("filtering pods on nodes, %w", err)
+			zap.S().Debugf("unable to get pods for node %s, %s", node.Name, err.Error())
+			continue
 		}
-		if f.isUnderutilized(&node, pods) {
-			underutilized = append(underutilized, &node)
+
+		// Only checks if it has 0 non-daemon pods right now
+		if f.isUnderutilized(pods) {
+			underutilized = append(underutilized, node)
 		}
 	}
 	return underutilized, nil
 }
 
+func (f *Filter) GetExpiredNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
+	nodes := []*v1.Node{}
+
+	nodeList, err := f.getNodes(ctx, provisioner)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, node := range nodeList {
+		if nodeutils.IsPastTTL(node) {
+			nodes = append(nodes, node)
+		}
+	}
+	return nodes, nil
+}
+
+func (f *Filter) getNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
+	nodes := &v1.NodeList{}
+	if err := f.kubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{
+		v1alpha1.ProvisionerNameLabelKey:      provisioner.Name,
+		v1alpha1.ProvisionerNamespaceLabelKey: provisioner.Namespace,
+	})); err != nil {
+		return nil, fmt.Errorf("listing nodes, %w", err)
+	}
+
+	// Convert each node to a pointer
+	nodePointers := []*v1.Node{}
```

Comment:
Why not just return []v1.Node instead of []*v1.Node?

Reply:
I had it originally to return `[]v1.Node` but then I resulted it resulted in that memory error. To further abstract for other functions that will use this helper function, I figured it would be best to return `[]*v1.Node` so that the mistake shouldn't be made again.

Reply:
That pointer gotcha is just something to be aware of in go. It's a noble goal to protect engineers from themselves, but there are many ways to construct this gotcha such that it's very hard to avoid in all cases. I'm not too particular on this detail, but my sense is that you could remove ~6 lines of code from this CR if you made it a []v1.Node.

Reply:
And to avoid that gotcha, you'd simple ptr.Node(node) instead of ptr.Node(*node)

Reply:
Thanks, I think I'm going to keep as is. If it becomes a bigger discussion, happy to refactor it later.